pkg/crc/machine/bundle/copier.go

```diff
@@ -0,0 +1,160 @@
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/code-ready/crc/pkg/compress"
+	"github.com/code-ready/crc/pkg/crc/logging"
+	crcos "github.com/code-ready/crc/pkg/os"
+	godeepcopy "github.com/jinzhu/copier"
+)
+
+type Copier struct {
+	srcBundle    *CrcBundleInfo
+	copiedBundle CrcBundleInfo
+}
+
+func NewCopier(srcBundle *CrcBundleInfo, basePath string, customBundleName string) (*Copier, error) {
+	var copier Copier
+
+	bundlePath := filepath.Join(basePath, customBundleName)
+	if err := os.Mkdir(bundlePath, 0775); err != nil {
+		return nil, err
+	}
+
+	// Creating a deepCopy of source bundle metadata so that whatever we make
+	// modification of any metadata field for copied operation should only
+	// effect the copied bundle metadata instead of source bundle metadata.
+	if err := godeepcopy.Copy(&copier.copiedBundle, srcBundle); err != nil {
+		return nil, err
+	}
+
+	copier.copiedBundle.bundleName = customBundleName
+	copier.copiedBundle.Type = "custom"
+	copier.srcBundle = srcBundle
+	copier.copiedBundle.cachedPath = bundlePath
+
+	return &copier, nil
+}
+
+func (copier *Copier) Cleanup() error {
+	return os.RemoveAll(copier.copiedBundle.cachedPath)
+}
+
+func (copier *Copier) resolvePath(filename string) string {
+	return copier.copiedBundle.resolvePath(filename)
+}
+
+func (copier *Copier) CachedPath() string {
+	return copier.copiedBundle.cachedPath
+}
+
+func (copier *Copier) CopyPrivateSSHKey(srcPath string) error {
+	sshKeyFileName := filepath.Base(copier.srcBundle.GetSSHKeyPath())
+	destPath := copier.resolvePath(sshKeyFileName)
+	return crcos.CopyFileContents(srcPath, destPath, 0400)
+}
+
+func (copier *Copier) SetKubeAdminPassword(kubePassword string) error {
+	kubeAdminPasswordFileName := copier.srcBundle.ClusterInfo.KubeadminPasswordFile
+	path := copier.resolvePath(kubeAdminPasswordFileName)
+	return ioutil.WriteFile(path, []byte(kubePassword), 0600)
+}
+
+func (copier *Copier) CopyKubeConfig() error {
+	kubeConfigFileName := filepath.Base(copier.srcBundle.GetKubeConfigPath())
+	srcPath := copier.srcBundle.GetKubeConfigPath()
+	destPath := copier.resolvePath(kubeConfigFileName)
+	return crcos.CopyFileContents(srcPath, destPath, 0640)
+}
+
+func (copier *Copier) CopyFilesFromFileList() error {
+	for _, file := range copier.srcBundle.Storage.Files {
+		srcPath := copier.srcBundle.resolvePath(file.Name)
+		destPath := copier.resolvePath(filepath.Base(srcPath))
+		if err := crcos.CopyFileContents(srcPath, destPath, 0755); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (copier *Copier) SetDiskImage(path string, format string) error {
+	fileInfo, err := getFileInfo(path)
+	if err != nil {
+		return err
+	}
+	diskImage := DiskImage{
+		File:   *fileInfo,
+		Format: format,
+	}
+
+	copier.copiedBundle.Storage.DiskImages = []DiskImage{diskImage}
+
+	return nil
+}
+
+func (copier *Copier) GenerateBundle(bundleName string) error {
```

Comment:
You could call `CopyFilesFromFileList()` unconditionally here instead of relying on the user of `Copier` to do it? I realize the same could be said for `CopyKubeConfig()`. But maybe better to keep things consistent and leave both out of the `GenerateBundle` ?

Reply:
yes I think it is better to keep those out of `GenerateBundle` so we have bit consistency in the flow (like copies all the required resource before call this function).

Reply:
This leaves a corner case as calling `CopyFilesFromFileList` is mandatory otherwise `Generate` will produce an invalid bundle, but that can be reworked later.